src/components/implementation/no_interface/new_boot/ppos_booter.c

```diff
@@ -0,0 +1,344 @@
+#include <stdio.h>
+#include <string.h>
+
+#include <cos_component.h>
+#include <cobj_format.h>
+#include <cos_kernel_api.h>
+#include <boot_deps.h>
+
+struct deps {
+	short int client, server;
+};
+struct deps *deps;
+int ndeps;
+
+static void
+cos_llprint(char *s, int len)
+{ call_cap(PRINT_CAP_TEMP, (int)s, len, 0, 0); }
+
+int
+prints(char *s)
+{
+		int len = strlen(s);
+		cos_llprint(s, len);
+		return len;
+}
+
+int __attribute__((format(printf,1,2)))
+printc(char *fmt, ...)
+{
+	  char s[128];
+  	  va_list arg_ptr;
+  	  int ret, len = 128;
+
+  	  va_start(arg_ptr, fmt);
+  	  ret = vsnprintf(s, len, fmt, arg_ptr);
+  	  va_end(arg_ptr);
+	  cos_llprint(s, ret);
+
+	  return ret;
+}
+
+/*Component init info*/
+#define INIT_STR_SZ 52
+struct component_init_str {
+        unsigned int spdid, schedid;
+        int startup;
+        char init_str[INIT_STR_SZ];
+}__attribute__((packed));
+
+struct component_init_str *init_args;
+
+unsigned int *boot_sched;
+
+static void
+boot_find_cobjs(struct cobj_header *h, int n)
+{
+	int i;
+	vaddr_t start, end;
+
+	start = (vaddr_t)h;
+	hs[0] = h;
+
+	for (i = 1 ; i < n ; i++) {
+		int j = 0, size = 0, tot = 0;
+
+		size = h->size;
+		for (j = 0 ; j < (int)h->nsect ; j++) {
+			tot += cobj_sect_size(h, j);
+		}
+		printc("cobj %s:%d found at %p:%x, size %x -> %x\n",
+		       h->name, h->id, hs[i-1], size, tot, cobj_sect_get(hs[i-1], 0)->vaddr);
+
+		end   = start + round_up_to_cacheline(size);
+		hs[i] = h = (struct cobj_header*)end;
+		start = end;
+	}
+	
+	hs[n] = NULL;
+	printc("cobj %s:%d found at %p -> %x\n",
+	       hs[n-1]->name, hs[n-1]->id, hs[n-1], cobj_sect_get(hs[n-1], 0)->vaddr);
+}
+
+static int
+boot_comp_map_memory(struct cobj_header *h, spdid_t spdid, pgtblcap_t pt)
+{
+	int i, j;
+	int flag;
+	vaddr_t dest_daddr, prev_map = 0;
+	int tot = 0, n_pte = 1;	
+	struct cobj_sect *sect = cobj_sect_get(h, 0);
+	
+	/* Expand Page table, could do this faster */
+	for (j = 0 ; j < (int)h->nsect ; j++) {
+		tot += cobj_sect_size(h, j);
+	}
+	
+	if (tot > SERVICE_SIZE) {
+		n_pte = tot / SERVICE_SIZE;
+		if (tot % SERVICE_SIZE) n_pte++;
+	}	
+	
+	boot_comp_pgtbl_expand(n_pte, pt, sect->vaddr);
+
+	/* We'll map the component into booter's heap. */
+	new_comp_cap_info[spdid].vaddr_mapped_in_booter = (vaddr_t)cos_get_heap_ptr();
+	
+	for (i = 0 ; i < h->nsect ; i++) {
+		int left;
+
+		sect = cobj_sect_get(h, i);
+		flag = MAPPING_RW;
+		if (sect->flags & COBJ_SECT_KMEM) {
+			flag |= MAPPING_KMEM;
+		}
+
+		dest_daddr = sect->vaddr;
+		left       = cobj_sect_size(h, i);
+		
+		/* previous section overlaps with this one, don't remap! */
+		if (round_to_page(dest_daddr) == prev_map) {
+			left -= (prev_map + PAGE_SIZE - dest_daddr);
+			dest_daddr = prev_map + PAGE_SIZE;
+		}
+
+		while (left > 0) {
+			boot_comp_mmap(spdid, dest_daddr);
```

Comment:
I love how this all feels very generic now.  You've done a great job abstracting away the `cos_kernel_api` stuff.